specification/dsdl/dsdl.tex

```diff
@@ -376,6 +431,108 @@ \section{Primitive data types}\label{sec:dsdl_primitive_data_types}
+                            \textbackslash+|-)?[0-9]*((e|E)-?[0-9]+) |
+                            \textbackslash+|-)?[0-9]*.[0-9]*((e|E)-?[0-9]+) } \\
+
+    \footnotesize \texttt{const char} &
+        \footnotesize $\lbrack0, 127\rbrack$ &
+        \footnotesize
+            \texttt{ '[\^{}'\textbackslash\textbackslash]' |
+                'x[0-9]+' |
+                '\textbackslash n' |
+                '\textbackslash r' |
+                '\textbackslash\textbackslash' |
+                '\textbackslash t' |
+                '\textbackslash ''} \\
+
+\end{UAVCANSimpleTable}
+
+Notice that the type are uniquely defined by the literals.
+This is done to avoid both explicitly noting const types and non-trivial type inference.
+
+
+\subsection{Array types}\label{sec:dsdl_array_data_types}
+DSDL contains a class of types that are somewhat in-between primitive and composite types, the arrays.
+The array types are similar to the primitive types that they are in fact built in to the DSDL language definition,
+on the other hand they're actually defined using primitive types.
+
+The class of array types in DSDL can be generated from two array kinds, the fixed size array and the dynamic array.
+Each of them needs a type (not restricted to primitive types)
+and a non-negative \texttt{const int} to define an array type.
+
+The type of an array with elements of type \verb|T| is denoted as the following:
+\begin{itemize}
+    \item Syntax \verb|T[S]| is a static array of size exactly S items.
+    \item Syntax \verb|T[<S]| is a dynamic array of size from 0 to S-1 items, inclusively.
+    \item Syntax \verb|T[<=S]| is a dynamic array of size from 0 to S items, inclusively.
+\end{itemize}
+
+In the array definition statements above, \verb|S| is a non-negative \verb|const int|.
+
+Observe that the maximum size of dynamic arrays is always bounded,
+this ensures that the worst case memory footprint and associated computational complexity are predictable.
+
+\subsection{Constant set types}
+The constant set kind generates homogenous set of values of constant type from a primitive constant type
+which the equality operation (equality checked with the \verb|==| operator) is defined for.
+If multiple items of the same value is inserted (related to the \verb|==| operator),
+it will have the same result as if only one item with that value was inserted.
+
+An instance of the constant set can be constructed by placing comma separated constant expressions inside braces.
+The element type is automatically resolved from the elements,
+making it currently impossible to express the empty set in a convenient way as
+there would be no way to know the type of \verb|{}|.
+
+Examples of legal literals to construct sets are the following:
+\begin{itemize}
+    \item \texttt{ \{ 1, 2, 3 \} }
+    \item \texttt{ \{ 1\%3, 2\%3, 3\%3 \} }
+    \item \texttt{ \{ 1, 2, 3 , 3, 3, 3, 3\} }
+    \item \texttt{ \{ 0x201, 0b2, -301 \} }
+    \item \texttt{ \{ true, false, true \} }
```

Comment:
What non-integer sets are good for? None of the defined operators seem to yield them.

Reply:
It just seemed more future proof to define sets on things that can be checked equivalence on rather than integers only. 

The only things you can do with sets of `const bool`s or `const char`s is to construct them, and you're not even allowed to construct sets of `const float`s. The things is that constructing a const set is not a legal statement (on its own, it must be assigned to something or used in an assertion) in DSDL, and there is no legal place they can be used. This means that any statement that constructs such set will be erroneous in DSDL even if the construction itself has meaning. 

I can make it clear in the text that constructing sets of other stuff is technically legal but there is no way it will not give an error down the line and must be avoided.

It's totally OK for a compiler to give an error already on the construction of these sets as they will never succeed one step down the line.

Reply:
I mean what's the point of defining something that cannot exist? Shall the need arise in the future, we can always add them later in a backward-compatible way ¯\\_(ツ)_/¯

Reply:
It's to give a reason to disallow constructing the empty set `{}` without any considerations to type. This must be disallowed if we're to add them later.

I will add a paragraph to disallow this instead.